src/libcollections/vec.rs

```diff
@@ -1652,16 +1521,16 @@ impl<T: Ord> Ord for Vec<T> {
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<T> Drop for Vec<T> {
     fn drop(&mut self) {
-        // This is (and should always remain) a no-op if the fields are
-        // zeroed (when moving out, because of #[unsafe_no_drop_flag]).
-        if self.cap != 0 && self.cap != mem::POST_DROP_USIZE {
-            unsafe {
-                for x in self.iter() {
-                    ptr::read(x);
-                }
-                dealloc(*self.ptr, self.cap)
+        // NOTE: this is currently abusing the fact that ZSTs can't impl Drop.
+        // Or rather, that impl'ing Drop makes them not zero-sized. This is
+        // OK because exactly when this stops being a valid assumption, we
+        // don't need unsafe_no_drop_flag shenanigans anymore.
+        if mem::size_of::<T>() != 0 && self.buf.cap() != mem::POST_DROP_USIZE {
```

Comment:
I'm pretty sure `cap != 0` was doubling as "drop flag version bridge" and
"don't deallocate Vec::new()". Neither of which apply now.

On Tue, Jul 14, 2015 at 3:59 PM, bluss <notifications@github.com> wrote:

> In src/libcollections/vec.rs
> <https://github.com/rust-lang/rust/pull/26955#discussion_r34630306>:
>
> > @@ -1652,16 +1521,16 @@ impl<T: Ord> Ord for Vec<T> {
> >  #[stable(feature = "rust1", since = "1.0.0")]
> >  impl<T> Drop for Vec<T> {
> >      fn drop(&mut self) {
> > -        // This is (and should always remain) a no-op if the fields are
> > -        // zeroed (when moving out, because of #[unsafe_no_drop_flag]).
> > -        if self.cap != 0 && self.cap != mem::POST_DROP_USIZE {
> > -            unsafe {
> > -                for x in self.iter() {
> > -                    ptr::read(x);
> > -                }
> > -                dealloc(*self.ptr, self.cap)
> > +        // NOTE: this is currently abusing the fact that ZSTs can't impl Drop.
> > +        // Or rather, that impl'ing Drop makes them not zero-sized. This is
> > +        // OK because exactly when this stops being a valid assumption, we
> > +        // don't need unsafe_no_drop_flag shenanigans anymore.
> > +        if mem::size_of::<T>() != 0 && self.buf.cap() != mem::POST_DROP_USIZE {
>
> Does this need to continue checking for the cap == 0 case? I'm not clear
> on if we ever zero out in the era of filling drop.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/26955/files#r34630306>.
>